pkg/reconciler/managed/reconciler.go

```diff
@@ -1248,3 +1291,92 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (resu
 	managed.SetConditions(xpv1.ReconcileSuccess())
 	return reconcile.Result{RequeueAfter: reconcileAfter}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 }
+
+// recordChangeLog constructs a change log entry from the provided information
+// and sends it to the change log service, if the feature is enabled. If errors
+// are encountered during this process, then logs/events will be recorded for
+// each, but the process will continue. We are choosing to send all the
+// information we can to the change logs as opposed to sending nothing.
+func (r *Reconciler) recordChangeLog(ctx context.Context, managed resource.Managed, snapshot *structpb.Struct, log logging.Logger,
+	record event.Recorder, opType protov1alpha1.OperationType, changeErr error, ad *AdditionalDetails,
+) {
+	if !r.features.Enabled(feature.EnableAlphaChangeLogs) {
+		// change log feature isn't enabled, just return immediately
+		return
+	}
+
+	// convert any additional details to a Struct to include in the change log entry
+	var ads *structpb.Struct
+	if ad != nil {
+		b, err := json.Marshal(ad)
+		if err != nil {
+			log.Debug("Cannot marshal additional details", "error", err)
+			record.Event(managed, event.Warning(reasonCannotSendChangeLog, err))
+		} else {
+			ads = &structpb.Struct{}
+			if err = ads.UnmarshalJSON(b); err != nil {
+				log.Debug("Cannot convert additional details to Struct", "error", err)
+				record.Event(managed, event.Warning(reasonCannotSendChangeLog, err))
+				ads = nil
+			}
+		}
+	}
+
+	// determine the full namespaced name of the managed resource
+	var namespacedName string
+	namespace := managed.GetNamespace()
+	if namespace != "" {
+		namespacedName = fmt.Sprintf("%s/%s", namespace, managed.GetName())
+	} else {
+		namespacedName = managed.GetName()
+	}
+
+	// get an error message from the error if it exists
+	var changeErrMessage *string
+	if changeErr != nil {
+		changeErrMessage = reference.ToPtrValue(changeErr.Error())
+	}
+
+	// send everything we've got to the change log service
+	cle := &protov1alpha1.ChangeLogEntry{
+		Provider:          r.providerVersion,
+		Type:              managed.GetObjectKind().GroupVersionKind().String(),
+		Name:              namespacedName,
+		ExternalName:      meta.GetExternalName(managed),
+		Operation:         opType,
+		Snapshot:          snapshot,
+		ErrorMessage:      changeErrMessage,
+		AdditionalDetails: ads,
+	}
+
+	clr, err := r.changeLogClient.SendChangeLog(ctx, cle)
```

Comment:
I'm considering adding the `WaitForReady()` call option here, so we automatically retry calls that fail due to a transient failure state. Some resources about using `WaitForReady()`:

* https://pkg.go.dev/google.golang.org/grpc#WaitForReady
* https://grpc.io/docs/guides/wait-for-ready/
* https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md

> By default, RPCs will fail if the ClientConn enters the "transient failure" state, but setting WaitForReady(true) on a call will cause it to queue even in the "transient failure" state, and it will only ever fail due to a deadline, a server response, or a connection loss after the RPC was sent to a server.

If we were to use `WaitForReady()`, we should probably have a new `context` here that is separate from the main `Reconcile()` context and has a shorter timeout/deadline.

I was curious what Crossplane does for the function pipeline gRPC calls:

* the main function pipeline looks to use a [round robin load balancer](https://github.com/crossplane/crossplane/blob/master/internal/xfn/function_runner.go#L229-L232) approach for all connections as part of the default service config
* the `render` CLI command appears to use [`waitForReady`](https://github.com/crossplane/crossplane/blob/master/cmd/crank/beta/render/render.go#L110-L112) for all connections as part of the default service config

I think it could make these calls more robust for dealing with transient failures if we are using the `WaitForReady()` call option with a reasonably short context deadline. any strong opinions here?

Reply:
@negz I went ahead and made this change in https://github.com/crossplane/crossplane-runtime/pull/754/commits/f7fb3a733a63a95ba0d4eb5f5555105d247dabd8. With the reading and prototyping I did with this, I like the behavior we get here. We'll wait/retry if the server (change logs sidecar) is temporarily unreachable for a default of 10s, and that value is configurable with a new `WithSendTimeout()` option. 

Let me know if you don't think this is a good approach, but with my testing I think it adds a bit more robustness for sending the change logs without making the `Reconcile()` hang for too long.